### ISSUE TITLE
_load_textdomain_just_in_time was called incorrectly

### ISSUE THREAD

**Issue by f14m07** [johnbillion/query-monitor#919]
Hi,
I have this message in my php error log since the last WP 6.7 update.
To see this error your site must be in a foreign language and you must have debug mode.

Notice: Function _load_textdomain_just_in_time was called <strong>incorrectly</strong>. Translation loading for the <code>query-monitor</code> domain was triggered too early. This is usually an indicator for some code in the plugin or theme running too early. Translations should be loaded at the <code>init</code> action or later. Please see <a href="https://developer.wordpress.org/advanced-administration/debug/debug-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 6.7.0.) in /Users/xxxx/Local Sites/clone-reve/app/public/wp-includes/functions.php on line 6114

This isn't the only plugin to have this problem.
You must be in debug mode and with a foreign lang site to reproduce this notice.
Could you update your code please ?

Thanks for all.

Have a nice day.

Frédéric

**Comment by johnbillion** [comment]
Thanks for the report. I've already been looking into a fix for this. It's ironically caused by the "Doing it wrong" handler in Query Monitor which gets triggered when another plugin loads a translation too early.

**Comment by f14m07** [comment]
It's true because having deactivated the other plugins that have the same error, I no longer get the query monitor error. Thanks for your quick response

**Comment by johnbillion** [comment]
@f14m07 Are you still seeing this problem with WordPress 6.7.1?

**Comment by f14m07** [comment]
No it's Ok Not new issue.

**Comment by cbratschi** [comment]
Still getting this error in WordPress 6.7.1:

[06-Dec-2024 16:46:04 UTC] PHP Notice:  Function _load_textdomain_just_in_time was called <strong>incorrectly</strong>. Translation loading for the <code>query-monitor</code> domain was triggered too early. This is usually an indicator for some code in the plugin or theme running too early. Translations should be loaded at the <code>init</code> action or later. Please see <a href="https://developer.wordpress.org/advanced-administration/debug/debug-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 6.7.0.) in /home/site/wp-includes/functions.php on line 6114

**Comment by felipeelia** [comment]
I'm seeing this problem in a different context now. ElasticPress makes a remote HTTP call using wp_remote_post() and as Query Monitor monitors all those, QM itself is calling __() earlier than expected [here](https://github.com/johnbillion/query-monitor/blob/develop/classes/Util.php#L200). I'm still debugging it but wanted to share this finding asap, so hopefully it helps. Also worth noting I'm having a hard time reproducing all these problems, as the notice will not be displayed unless I delete all transients using `wp transient delete --all`.

**Comment by johnbillion** [comment]
Thanks. This seems like a never ending problem once you dig into it.

I'll need to do some work to shift all the translation strings out of the collectors and into the output.

**Comment by afragen** [comment]
Just documenting a conversation with John. I was seeing similar errors and think it may be do to a response from `wp_remote_get()` returning something that QM is trying to translate from `doing_it_wrong`. I believe the following lines from my log are the inciting translations from QM.

```
[18-Dec-2024 20:04:58 UTC] PHP  17. __($text = 'Plugin: %s', $domain = 'query-monitor') /Users/afragen/Local_Sites/lightning/app/public/wp-content/plugins/query-monitor/classes/Util.php:200
[18-Dec-2024 20:04:58 UTC] PHP  18. translate($text = 'Plugin: %s', $domain = 'query-monitor') /

[18-Dec-2024 20:04:58 UTC] PHP  37. __($text = 'WordPress Core', $domain = 'query-monitor') /Users/afragen/Local_Sites/lightning/app/public/wp-content/plugins/query-monitor/classes/Util.php:246
[18-Dec-2024 20:04:58 UTC] PHP  38. translate($text = 'WordPress Core', $domain = 'query-monitor') /
```

What a PITA this change has been. Silently failing seems to be so much better.

**Comment by afragen** [comment]
Got tired of seeing this and with some help from John (thanks John) I give you.

https://gist.github.com/afragen/986ba6046ebea963074283a77cae07fc

**Comment by kagg-design** [comment]
It can be temporarily fixed by #940.

**Comment by ju1ius** [comment]
I just upgraded a site to WP 6.7.1 (w/ QM 3.17.0) and this triggered an infinite recursion loop ending in a BSOD (interestingly PHP OOMed before reaching max recursion depth).

I step-debugged the issue and here's the call graph (with irrelevant frames removed):
```
do_action('plugins_loaded')
┗ [WPML] current_user_can('manage_options')
  ┗ apply_filters('user_has_cap', ...)
    ┗ QM_Collector_Caps::filter_user_has_cap()
      ┗ QM_Backtrace::get_component()
        ┗ QM_Util::get_file_component()
          ┗ translate('Plugin: %s', 'query-monitor')
            ┗ _load_textdomain_just_in_time('query-monitor')
              ┗ _doing_it_wrong('translation loaded too early ...')
                ┗ QM_Collector_Doing_It_Wrong::maybe_prevent_error(true)
           ┏━━━━▷ ┗ current_user_can('view_query_monitor')
           ┃        ┗ QueryMonitor::filter_user_has_cap()
           ┃          ┗ user_can($user, 'manage_options')
           ┃            ┗ QM_Collector_Caps::filter_user_has_cap()
           ┃              ┗ QM_Backtrace::get_component()
           ┃                ┗ QM_Util::get_file_component()
           ┃                  ┗ translate('WordPress Core', 'query-monitor')
           ┃                    ┗ _load_textdomain_just_in_time('query-monitor')
           ┃                      ┗ _doing_it_wrong('translation loaded too early ...')
           ┃                        ┗ QM_Collector_Doing_It_Wrong::maybe_prevent_error(true)
           ┗━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

> I'll need to do some work to shift all the translation strings out of the collectors and into the output.

@johnbillion It seems to be the only proper solution... 🙏🏻

**Comment by johnbillion** [comment]
Yeah I agree. I'll find some time this week to look into at least a temporary means of avoiding this and get a release out.
